src/Invio.Extensions.Core/Collections/IListExtensions.cs

```diff
@@ -206,5 +215,35 @@ public static class IListExtensions {
             v5 = source[4];
             rest = source.Skip(5);
         }
+
+        /// <summary>
+        /// Shuffles a list in place by iterating through each item and randomly
+        /// picking a target item to swap places with.
+        /// </summary>
+        /// <remarks>
+        /// This implementation does not guarantee that any items will actually
+        /// change their current locations in the list.
+        /// </remarks>
+        /// <param name="source">
+        /// The list to shuffle.</param>
+        /// <typeparam name="T">The type contained by the list.</typeparam>
+        /// <exception cref="ArgumentNullException">The source list was null.</exception>
+        public static void Shuffle<T>(this IList<T> source) {
+            if (source == null) {
+                throw new ArgumentNullException(nameof(source));
+            }
+
+            for (var sourceIndex = 0; sourceIndex < source.Count; sourceIndex++) {
+                var targetIndex = random.Value.Next(0, source.Count);
+
+                if (sourceIndex != targetIndex) {
+                    var tmp = source[sourceIndex];
+                    source[sourceIndex] = source[targetIndex];
+                    source[targetIndex] = tmp;
+                }
+            }
+        }
```

Comment:
It turns out that this algorithm doesn't actually generate a normally distributed set of shuffled lists. It actually has a small bias towards certain orderings.

For example consider the input "abc". Three calls to `random.Value.Next(0, source.Count)` should generate with equal probability one of the following sequences of values:

```json
[ [0,0,0],
  [0,0,1],
  [0,0,2],
  [0,1,0],
  [0,1,1],
  [0,1,2],
  [0,2,0],
  [0,2,1],
  [0,2,2],
  [1,0,0],
  [1,0,1],
  [1,0,2],
  [1,1,0],
  [1,1,1],
  [1,1,2],
  [1,2,0],
  [1,2,1],
  [1,2,2],
  [2,0,0],
  [2,0,1],
  [2,0,2],
  [2,1,0],
  [2,1,1],
  [2,1,2],
  [2,2,0],
  [2,2,1],
  [2,2,2] ]
```

Each of these sequences generates the following 27 possible outputs:

```json
"cab",
"bca",
"bac",
"cba",
"acb",
"abc",
"bca",
"abc",
"acb",
"cba",
"acb",
"abc",
"cab",
"bca",
"bac",
"acb",
"bac",
"bca",
"acb",
"bac",
"bca",
"abc",
"cab",
"cba",
"bac",
"cba",
"cab"
```

If we count these we get the following:

```json
{
  "abc": 4,
  "acb": 5,
  "bac": 5,
  "bca": 5,
  "cab": 4,
  "cba": 4
}
```

I'm pretty sure that if the low limit for your random number is `sourceIndex` instead of `0` you will get a normal distribution, and you can also skip the `sourceIndex == source.Count - 1` case since at that point `sourceIndex` will always equal `targetIndex`.

In my example of "abc" before, the possible lists of swap indexes given this algorithm would be:

```javascript
[ [0, 1], // abc 
  [1, 1], // bac
  [2, 1], // cba
  [0, 2], // acb
  [1, 2], // bca
  [2, 2]  // cab
]
```

So we should get each possible shuffle with an equal probability. Y'know, just in case we pivot to making an online poker site.

Reply:
Le sigh. Resolving in #4